src/cosalib/ostree.py

```diff
@@ -0,0 +1,221 @@
+            'blocks': blks_regfiles},
+        'symlinks': {
+            'count': n_symlinks,
+            'blocks': blks_symlinks
+        },
+        'inodes': {
+            'count': n_inodes,
+            'mb': total_inode_mb,
+        },
+        'estimate-mb': {
+            'base': total_mb,
+            'final': estimate_mb
+        },
+    }
+
+
+class OSTreeException(Exception):
+    pass
+
+
+class _BuildOSTree():
+    """
+    BuildOSTree interacts with OS Trees written to <dir>/tmp/repo from the
+    CoreOS Assembler command "cosa build."
+
+    The core consumer is cosalib.metal.MetalVariant.
+    """
+    def __init__(self, build):
+        self._build = build
+        if not isinstance(build, _Build):
+            raise OSTreeException("BuildOStree expects a _Build Object")
+
+        self._ostree_repo = os.path.join(build.workdir, "tmp", "repo")
+        if not os.path.exists(self._ostree_repo):
+            raise OSTreeException(
+                f"missing ostree repo at {self._ostree_repo}")
+
+        # find the compose-*json file to discern the ref we expect
+        try:
+            os.chdir(os.path.join(build.workdir, "tmp"))
+            compose_candidates = glob.glob("compose-*.json")
+            if len(compose_candidates) > 1:
+                raise Exception("found multiple compose json files")
+            c_file = compose_candidates[0]
+            log.info(f"Reading {os.getcwd()}/{c_file} for compose information")
+            with open(c_file, 'r') as data:
+                self._compose_data = json.load(data)
+            log.info(f"Using OSTree ref '{self.ref_name}'")
+        finally:
+            os.chdir(build.workdir)
+
+    @property
+    def build(self):
+        return self._build
+
+    @property
+    def ostree_repo(self):
+        return self._ostree_repo
+
+    @property
+    def ref_name(self):
+        return self._compose_data.get("ref",
+                                      f"tmpref-{self.build.build_name}")
+
+    @property
+    def ostree_repo_size(self):
+        ors = getattr(self, "_ostree_repo_size", None)
+        if ors is None:
+            self._ostree_repo_size = estimate_tree_size(
+                self.ostree_repo, self.ref_name)
+        return self._ostree_repo_size
+
+    @property
+    def ostree_repo_estimated_size(self):
+        return self.ostree_repo_size.get("estimated-mb", {}).get("final", 4096)
+
+    @property
+    def ostree_tarball_meta(self):
+        return self.meta.get("images", {}).get("ostree", {})
+
+    @property
+    def ostree_tarball_path(self):
+        return self.ostree_meta.get("path")
+
+    @property
+    def ostree_tarball_sha256(self):
+        return self.ostree_meta.get("sha256")
+
+    @property
+    def ostree_size(self):
+        return self.ostree_meta.get("size")
+
+    def need_tree(self):
+        log.info("Checking for OSTree")
+        repo_path = self.ostree_repo
+        create = False
+        try:
+            _ = self.ostree_repo
```

Comment:
Having property access throw exceptions feels like a bad idea to me.  (I'd be surprised if there wasn't a Python style guide recommending against this)  If it's going to do something nontrivial best make it a function call.

Reply:
[pep008](https://www.python.org/dev/peps/pep-0008/#designing-for-inheritance) does denote there _generally_ should be no other side effects from properties other than caching. Throwing an `AttributeError` may be OK since that is what would happen if a variable isn't set.

In this case it looks like the exception is caught and handled. There _is_ a side effect of "setting up" which seems fair.